### PR TITLE
Upgrade undici and update deps

### PR DESCRIPTION
**Pull request by simone-sanfratello**
Upgrade undici to `v4`

Note the proposal of drop the node `v10` support because `undici` is not supporting it

~~typecheck fails due to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53820, so I'm also waiting for that~~

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by simone-sanfratello** [comment]
on node `v10` is failing because of `queueMicrotask`

https://github.com/nodejs/undici/blob/main/lib/api/api-request.js#L135

I see `undici` is not supporting node `v10` so I guess the consistent solution is to drop node `v10` on `fastify-reply-from` too

**Comment by kibertoad** [comment]
@simone-sanfratello `fastify` itself supports down to Node 10.16.0, so I don't think that we can remove support for 10.x until the next fastify semver major.

**Comment by simone-sanfratello** [comment]
we have 2 options, either add node `v10` support on `undici` or drop node `v10` in `fastify-reply-from`

I'm proposing this PR for the second solution

**Comment by kibertoad** [comment]
@simone-sanfratello What I'm trying to say, it's going to be confusing for users if Node requirements for one of the plugins are stricter than the requirements for fastify itself, and it will block all future updates to the plugin for everyone on older Node versions. This may prevent this PR from landing in the nearby future.

**Comment by mcollina** [comment]
+1 to drop support for Node 10.

**Comment by kibertoad** [comment]
@mcollina Should there be an explicit text in readme stating that only Node 12+ is supported and users of older node should use older version of the plugin, ir just the engine entry is sufficient?

File: package.json [hunk 1]
```diff
@@ -27,39 +27,39 @@
...
+    "end-of-stream": "^1.4.4",
     "fastify-plugin": "^3.0.0",
     "http-errors": "^1.8.0",
     "pump": "^3.0.0",
-    "semver": "^7.2.1",
-    "tiny-lru": "^7.0.0",
-    "undici": "^3.3.1"
+    "semver": "^7.3.5",
+    "tiny-lru": "^7.0.6",
+    "undici": "^4.0.0-rc.7"
```
Comment (kibertoad) [inline package.json:62]: 4.0.0 has already landed

**Comment by simone-sanfratello** [comment]
> @mcollina Should there be an explicit text in readme stating that only Node 12+ is supported and users of older node should use older version of the plugin, or just the engine entry is sufficient?

imo this should be a major release, but happy yo add some lines in the readme to make it clear

**Comment by mcollina** [comment]
Yes, it's a semver-major change.

File: lib/request.js [hunk 2]
```diff
@@ -120,7 +120,9 @@ function buildRequest (opts) {
       done(new Error('unix socket not supported with undici yet'))
       return
     } else {
-      pool = undiciAgent.get(baseUrl || opts.url.origin)
+      const options = getUndiciOptions(opts.undici)
+      options.dispatcher = undiciAgent
+      pool = new undici.Client(baseUrl || opts.url.origin, options)
```
Comment (mcollina) [inline lib/request.js:125]: This should be a Pool, not a client.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/request.js [hunk 3]
```diff
@@ -120,7 +120,9 @@ function buildRequest (opts) {
       done(new Error('unix socket not supported with undici yet'))
       return
     } else {
-      pool = undiciAgent.get(baseUrl || opts.url.origin)
+      const options = getUndiciOptions(opts.undici)
+      options.dispatcher = undiciAgent
```
Comment (mcollina) [inline lib/request.js:124]: are you sure this is needed?

Comment (simone-sanfratello) [reply lib/request.js:124]: I think so, because of the previous instruction `pool = undiciAgent.get`

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Fdawgs — APPROVED** [review]
👌🏻 Good job!

**Pull request merged**